refactor ducktape redpanda service in prep for k8s

andrewhsu opened the pull request:
Fixes #9687

This PR adds a `RedpandaServiceBase` class for `RedpandaService` to inherit from. This is a stepping-stone refactor in preparation for addressing issue #9709. A follow-up PR will add a class like `RedpandaK8sService` defined that inherits from `RedpandaServiceBase` and overrides the methods required to operate on a k8s deployment.

There should be no functionality change in this PR in regards to how the tests are run today.

To limit the scope of this refactor, i've only added the public methods In the `RedpandaServiceBase` class that need override based on what is used by the tests in PR https://github.com/redpanda-data/redpanda/pull/9535

The new `RedpandaServiceBase` class has these methods that need to be overridden in its sublcass:
* start_node
* stop_node
* clean_node
* get_node_memory_mb
* get_node_cpu_count
* get_node_disk_free
* lsof_node
* node_id
* set_cluster_config

And these methods look to be generic enough to be moved from `RedpandaService` into `RedpandaServiceBase` class:
* set_extra_rp_conf
* set_si_settings
* add_extra_rp_conf
* metrics
* metric_sum
* healthy
* partitions
* restart_nodes
* rolling_restart_nodes
* set_resource_settings

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.1.x
- [ ] v22.3.x
- [ ] v22.2.x

## Release Notes

* none

dlex commented on tests/rptest/services/redpanda.py:
```diff
@@ -960,6 +960,56 @@ def signal_redpanda(self, node, signal=signal.SIGKILL, idempotent=False):
+        """
+        pass
+
+    def rolling_restart_nodes(self,
```
This and some other methods (e.g. `metric_sum()`) seem like a higher level algorithms to me that do not need overrides in containerized/k8ized descendants, but rather an implementation in the base class using other virtual methods like `start_node()` or `metrics_sample()`.

IMO the most feasible approach to the DT/k8s adoption would be to select the simplest of all the tests in scope, and begin with porting that test only, rather than trying to cover them all at once. That would likely give the most basic abstraction required. Expanding from there to more and more tests will make it clear which methods require overloading and which do not.

andrewhsu replied:
looks like this method makes use of [RollingRestarter](https://github.com/redpanda-data/redpanda/blob/dev/tests/rptest/services/rolling_restarter.py#L15) class which is tightly-coupled with the `RedpandaService`. i'll work out what the deps are and bring them into this base class.

i've created a follow-on issue #9709 to track implementation of the k8s-specific methods which i'll get to after this PR merges and the k8s test infrastructure is up (tracked in https://github.com/redpanda-data/devprod/issues/650)

andrewhsu replied:
i've moved this impl into the base class with d17853eb247a65533e530548c9d511a0ffd7a576

andrewhsu replied:
update: i've force-pushed and this is now 409004d4b28e150a083c847f5bbba7bd0aa4e9f0

jcsp commented on tests/rptest/services/redpanda.py:
```diff
@@ -736,7 +736,8 @@ def __init__(self,
                  schema_registry_config: Optional[SchemaRegistryConfig] = None,
                  disable_cloud_storage_diagnostics=False):
-        super(RedpandaService, self).__init__(context, num_nodes=num_brokers)
+        super(RedpandaServiceBase, self).__init__(context,
```
I think you'd need to separate out this constructor into a base part and a RedpandaService part: e.g. `environment` is probably not going to be usable on kubernetes where we can set redpanda's environment variables, `superuser` may also not be settable, etc.

andrewhsu replied:
i've reduced the parameters in the constructor of `RedpandaServiceBase` to only include what is used in the base class instead of taking the entire signature from the original `RedpandaService` class.

jcsp commented on tests/rptest/services/redpanda.py:
```diff
@@ -832,6 +833,188 @@ def __init__(self,
+        nodes=None,
+        metrics_endpoint: MetricsEndpoint = MetricsEndpoint.METRICS,
+    ) -> Optional[MetricSamples]:
+        """
```
Let's not duplicate docstrings, they'll get stale.  Either omit them on the base class or the derived class (not strongly opinionated on which)

andrewhsu replied:
i'll remove them from the base class so the diff is smaller

jcsp commented on tests/rptest/services/redpanda.py:
```diff
@@ -832,6 +833,188 @@ def __init__(self,
+        """
+        pass
+
+    def signal_redpanda(self, node, signal=signal.SIGKILL, idempotent=False):
```
I'm not sure signal is really the portable primitive here: we usually want to stop redpanda (either hard or soft), or to do something funky like SIGSTOP'ing (pausing) the process.

For the stopping cases, this should be a stop_node thing so that it can do whatever kubernetes business is needed (like tainting a node so that it doesn't immediately restart), and for the SIGSTOP case that maybe needs to be some explicit method that can be clearly not-implemented on some backends.

andrewhsu replied:
i've removed `signal_redpanda` from the base class with 20698a24773acd79825a312d6a642af447ace67e because it looks like `stop_node` has a `forced` arg that is implemented in `RedpandaService` as `SIGKILL`

jcsp commented on tests/rptest/services/redpanda.py:
```diff
@@ -832,6 +833,188 @@ def __init__(self,
+        """
+        pass
+
+    def partitions(self, topic):
```
Wouldn't this method be the same for any backend?  It's just using Kafka API

andrewhsu replied:
i've moved this impl into the base class with 3466c44baa2cb59e8172f8c6b900209b7a7cece0

andrewhsu replied:
update: i've force-pushed and this is now 8cbffed5285c76bf888d5c15fcd9eae451231119

jcsp reviewed (commented):
A few comments inline.  I don't necessarily object to merging this ahead of your following changes, but agree with Alexey that we probably won't know if this is the right interface until the first alternative backend is working.

andrewhsu commented:
> A few comments inline. I don't necessarily object to merging this ahead of your following changes, but agree with Alexey that we probably won't know if this is the right interface until the first alternative backend is working.

i've reduced the scope of this PR to just the methods used in PR #9535 tests. see the updated PR description for the list of methods.

i've also rebased with latest on `dev` and cleaned up the commits. ready for review.

dlex commented on tests/rptest/services/redpanda.py:
```diff
@@ -726,6 +726,9 @@ def get_node_cpu_count(self):
 
+    def lsof_node(self, node: ClusterNode, filter: Optional[str] = None):
+        pass
+
```
nit: can be squashed with prev commit; actually all "add method to ovverride" commits can be squashed together

andrewhsu replied:
sorry, i didn't catch this comment because i went straight for the merge button. thanks for the review! :twisted_rightwards_arrows:

dlex reviewed (approved):
lgtm!
